cabby/model/landmark_recognition/dataset.py

```diff
@@ -0,0 +1,179 @@
+  `labels`: The ground true label of the cellid.
+  `cellids`: The ground truth S2Cell id.
+  `neighbor_cells`: One neighbor cell id of the ground truth S2Cell id.
+  `far_cells`: One far away cell id (in the region defined) of the ground truth
+  'dprob': Gamma distribution probability.
+  S2Cell id.
+  """
+
+  def __init__(self, data: pd.DataFrame):
+    # Tokenize instructions.
+
+    self.inputs = [
+      tokenize_and_preserve_labels(sent, labs)
+      for sent, labs in zip(data.instructions.tolist(), data.entity_span)
+    ]
+    self.sent = data.instructions.tolist()
+
+
+  def __getitem__(self, idx: int):
+    '''Supports indexing such that TextGeoDataset[i] can be used to get
+    i-th sample.
+    Arguments:
+      idx: The index for which a sample from the dataset will be returned.
+    Returns:
+      A single sample including text, the correct cellid, a neighbor cellid,
+      a far cellid, a point of the cellid and the label of the cellid.
+    '''
+
+    input = {k: torch.tensor(v) for k, v in self.inputs[idx].items()}
+    input['instructions'] = self.sent[idx]
+    return input
+
+  def __len__(self):
+    return len(self.inputs)
+
+
+def tokenize_and_preserve_labels(sentence, text_labels):
+  labels = []
+  spans = sorted(list(text_labels.values()))
+
+  sentence_words = word_tokenize(sentence)
+  cur_index = 0
+  span = spans.pop(0)
+  start, end = span[0], span[1]
+  for word in sentence_words:
+    if cur_index >= start:
+      labels.append(1)
+    else:
+      labels.append(0)
+
+    cur_index += len(word)
+    if cur_index<len(sentence) and sentence[cur_index] == ' ':
+      cur_index+=1
+    if cur_index >= end:
+      if len(spans)>0:
+        span = spans.pop(0)
+        start, end = span[0], span[1]
+      else:
+        start = len(sentence)+1
+
+  tokenized_input = tokenize_and_align_labels(labels, sentence_words)
+  return tokenized_input
+
+def tokenize_and_align_labels(tags, sentence_word, label_all_tokens = True):
+  tokenized_inputs = tokenizer(sentence_word, truncation=True, is_split_into_words=True)
+  word_ids = tokenized_inputs.word_ids(batch_index=0)
+  previous_word_idx = None
+  label_ids = []
+  for word_idx in word_ids:
+    # Special tokens have a word id that is None. We set the label to -100 so they are automatically
+    # ignored in the loss function.
+    if word_idx is None:
+      label_ids.append(-100)
+    # We set the label for the first token of each word.
+    elif word_idx != previous_word_idx:
+      label_ids.append(tags[word_idx])
+    # For the other tokens in a word, we set the label to either the current label or -100, depending on
+    # the label_all_tokens flag.
+    else:
+      label_ids.append(tags[word_idx] if label_all_tokens else -100)
+    previous_word_idx = word_idx
+
+  assert len(label_ids) == len(tokenized_inputs['input_ids'])
+  tokenized_inputs["labels"] = label_ids
+  return tokenized_inputs
+
+
+
+class PadSequence:

```

Comment:
Why is this a class and not just a function pad_sequence(batch)?

Reply:
This is the Pytorch way.

Reply:
Okay!